src/fpm_targets.f90

```diff
@@ -334,6 +336,57 @@ end function get_object_name
 end subroutine build_target_list
 
 
+!> Add non-library non-module dependencies for executable targets
+!>
+!>  Executable targets will link to any non-program non-module source files that
+!>   are in the same directory or in a subdirectory.
+!>
+!>  (Note: Fortran module dependencies are handled separately in
+!>    `resolve_module_dependencies` and `resolve_target_linking`.)
+!>
+subroutine collect_exe_link_dependencies(targets)
+    type(build_target_ptr), intent(inout) :: targets(:)
+
+    integer :: i, j
+    character(:), allocatable :: exe_source_dir
+
+    ! Add non-module dependencies for executables
+    do j=1,size(targets)
+
+        if (targets(j)%ptr%target_type == FPM_TARGET_EXECUTABLE) then
+
+            do i=1,size(targets)
+
+                if (i == j) cycle
+
+                associate(exe => targets(j)%ptr, dep => targets(i)%ptr)
+
+                    exe_source_dir = dirname(exe%dependencies(1)%ptr%source%file_name)
+
+                    if (allocated(dep%source)) then
+
+                        if (dep%source%unit_scope /= FPM_SCOPE_LIB .and. &
+                            dep%source%unit_type /= FPM_UNIT_PROGRAM .and. &
+                            dep%source%unit_type /= FPM_UNIT_MODULE .and. &
+                            index(dirname(dep%source%file_name), exe_source_dir) == 1) then
```

Comment:
````fortran
subroutine a
end subroutine a
module b_m
end module b_m
````
It seems that we still can't handle the case of Fortran source files mixed with `non-module procedure` and `module`? However, this mixed writing is not a good coding style, and we can ignore it for now.

Reply:
Actually that case should still work. If a source file contains a non-module procedure then it is marked as `FPM_UNIT_SUBPROGRAM` not `FPM_UNIT_MODULE` as per the rules needed for tree-shaking, see [here](https://fortran-lang.github.io/fpm/module/fpm_model.html).